coffea/nanoevents/transforms.py

```diff
@@ -288,6 +288,138 @@ def children(stack):
     stack.append(out)
 
 
+@numba.njit
+def _distinctChildrenDeep_kernel(offsets_in, global_parents, global_pdgs):
+    offsets_out = numpy.empty(len(global_parents) + 1, dtype=numpy.int64)
+    content_out = numpy.empty(len(global_parents), dtype=numpy.int64)
+    offsets_out[0] = 0
+
+    offset0 = 1
+    offset1 = 0
+    for record_index in range(len(offsets_in) - 1):
+        start_src, stop_src = offsets_in[record_index], offsets_in[record_index + 1]
+
+        for index in range(start_src, stop_src):
+            this_pdg = global_pdgs[index]
+
+            # only perform the deep lookup when this particle is not already part of a decay chain
+            # otherwise, the same child indices would be repeated for every parent in the chain
+            # which would require content_out to have a length that isa-priori unknown
+            if (
+                global_parents[index] >= 0
+                and this_pdg != global_pdgs[global_parents[index]]
+            ):
+                # keep an index of parents with same pdg id
+                parents = numpy.empty(stop_src - index, dtype=numpy.int64)
```

Comment:
That might be possible. The upper bound should be the length or particles itself, and then one could search between 2 offsets. I'll give it a try.

Reply:
Thinking about it, the `parents_with_children` array should be moved out of the loops as well, but I'm anyway not 100% happy with how children with same pdg id but no with further grandchildren (*) are identified. Do you think it's a good idea to pass the children index to the kernel to identify (*)? This could be used here:

```python
# then, depending on the pdg id, add to parents or content
if possibe_child_pdg == this_pdg and has_children(possible_child):
    # has the same pdg id and no children itself, add to parents
    ...
```

Reply:
As far as I know, numba should be able to compile functions that have local functions defined inside. So yeah that would work and probably improve readability